drivers/can/can_sccan.c

```diff
@@ -0,0 +1,1181 @@
+ */
+
+#define DT_DRV_COMPAT sc_can
+
+#include <zephyr/drivers/can.h>
+#include <zephyr/drivers/can/transceiver.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/byteorder.h>
+
+LOG_MODULE_REGISTER(sc_can, CONFIG_CAN_LOG_LEVEL);
+
+/* Registers */
+#define SCCAN_ENR_OFFSET    (0x0000) /* CAN Enable Register */
+#define SCCAN_TQPR_OFFSET   (0x0008) /* CAN Time Quantum Prescaler Register */
+#define SCCAN_BTSR_OFFSET   (0x000C) /* CAN Bit Timing Setting Register */
+#define SCCAN_ECNTR_OFFSET  (0x0010) /* CAN Error Count Register */
+#define SCCAN_STSR_OFFSET   (0x0018) /* CAN Status Register */
+#define SCCAN_FIFOSR_OFFSET (0x001C) /* CAN FIFO Status Register */
+#define SCCAN_ISR_OFFSET    (0x0020) /* CAN Interrupt Status Register */
+#define SCCAN_IER_OFFSET    (0x0024) /* CAN Interrupt Enable Register */
+#define SCCAN_TMR1_OFFSET   (0x0030) /* CAN TX Message Register1 */
+#define SCCAN_TMR2_OFFSET   (0x0034) /* CAN TX Message Register2 */
+#define SCCAN_TMR3_OFFSET   (0x0038) /* CAN TX Message Register3 */
+#define SCCAN_TMR4_OFFSET   (0x003C) /* CAN TX Message Register4 */
+#define SCCAN_THPMR1_OFFSET (0x0040) /* CAN TX High Priority Message Register1 */
+#define SCCAN_THPMR2_OFFSET (0x0044) /* CAN TX High Priority Message Register2 */
+#define SCCAN_THPMR3_OFFSET (0x0048) /* CAN TX High Priority Message Register3 */
+#define SCCAN_THPMR4_OFFSET (0x004C) /* CAN TX High Priority Message Register4 */
+#define SCCAN_RMR1_OFFSET   (0x0050) /* CAN RX Message Register1 */
+#define SCCAN_RMR2_OFFSET   (0x0054) /* CAN RX Message Register2 */
+#define SCCAN_RMR3_OFFSET   (0x0058) /* CAN RX Message Register3 */
+#define SCCAN_RMR4_OFFSET   (0x005C) /* CAN RX Message Register4 */
+#define SCCAN_AFER_OFFSET   (0x0060) /* CAN Acceptance Filter Enable Register */
+#define SCCAN_AFIMR1_OFFSET (0x0070) /* CAN Acceptance Filter ID Mask Register1 */
+#define SCCAN_AFIVR1_OFFSET (0x0074) /* CAN Acceptance Filter ID Value Register1 */
+#define SCCAN_AFIMR2_OFFSET (0x0090) /* CAN Acceptance Filter ID Mask Register2 */
+#define SCCAN_AFIVR2_OFFSET (0x0094) /* CAN Acceptance Filter ID Value Register2 */
+#define SCCAN_AFIMR3_OFFSET (0x00B0) /* CAN Acceptance Filter ID Mask Register3 */
+#define SCCAN_AFIVR3_OFFSET (0x00B4) /* CAN Acceptance Filter ID Value Register3 */
+#define SCCAN_AFIMR4_OFFSET (0x00D0) /* CAN Acceptance Filter ID Mask Register4 */
+#define SCCAN_AFIVR4_OFFSET (0x00D4) /* CAN Acceptance Filter ID Value Register4 */
+#define SCCAN_FIFORR_OFFSET (0x00F0) /* CAN FIFO and Buffer Reset Register */
+#define SCCAN_STMCR_OFFSET  (0x0100) /* CAN Self Test Mode Control Register */
+#define SCCAN_PSLMCR_OFFSET (0x0200) /* CAN PHY Sleep Mode Control Register */
+#define SCCAN_VER_OFFSET    (0xF000) /* CAN Controller IP Version Register */
+
+/* CAN Enable Register */
+#define SCCAN_CANEN BIT(0)
+
+/* CAN Error Count Register */
+#define SCCAN_RXECNT(x) (((x) & GENMASK(15, 8)) >> 8)
+#define SCCAN_TXECNT(x) (((x) & GENMASK(7, 0)))
```

Comment:
```suggestion
#define SCCAN_RX_ECR(x) (((x) & GENMASK(15, 8)) >> 8)
#define SCCAN_TX_ECR(x) (((x) & GENMASK(7, 0)))
```
ECR for Error counter register improve readability in my opinion.